PIA port forwarding - Bad Gateway

For a few days now I am no longer able to get port forwarding working. I have not changed anything, but am seeing this now in the logs:
`2021/04/16 00:09:14 ERROR port forwarding: cannot obtain token: 502 Bad Gateway: response received: "<html>\r<head><title>502 Bad Gateway</title></head>\r<body>\r<center><h1>502 Bad Gateway</h1></center>\r<hr><center>nginx/1.19.1</center>\r</body>\r</html>\r"`

Hmm, looks like there might be an issue:

```
=========================================,
================ Gluetun ================,
=========================================,
==== A mix of OpenVPN, DNS over TLS, ====,
======= Shadowsocks and HTTP proxy ======,
========= all glued up with Go ==========,
=========================================,
=========== For tunneling to ============,
======== your favorite VPN server =======,
=========================================,
=== Made with ❤️  by github.com/qdm12 ====,
=========================================,
,
Running version fix-pia-token built on 2021-04-17T13:48:42Z (commit 12dca86),
,
,
🔧  Need help? https://github.com/qdm12/gluetun/issues/new,
💻  Email? quentin.mcgaw@gmail.com,
☕  Slack? Join from the Slack button on Github,
💸  Help me? https://github.com/sponsors/qdm12,
2021/04/17 10:24:31 INFO Unbound version: 1.10.1,
2021/04/17 10:24:31 INFO IPtables version: v1.8.4,
2021/04/17 10:24:31 INFO OpenVPN version: 2.4.10,
2021/04/17 10:24:31 INFO Settings summary below:,
|--OpenVPN:,
   |--Verbosity level: 1,
   |--Run as root: enabled,
   |--Provider:,
      |--Private Internet Access settings:,
         |--Network protocol: udp,
         |--Regions: ca toronto,
         |--Encryption preset: normal,
         |--Custom port: 0,
         |--Port forwarding:,
            |--File path: /tmp/gluetun/forwarded_port,
|--DNS:,
   |--Plaintext address: 1.1.1.1,
   |--DNS over TLS:,
      |--Unbound:,
          |--DNS over TLS providers:,
              |--cloudflare,
          |--Listening port: 53,
          |--Access control:,
              |--Allowed:,
                  |--0.0.0.0/0,
                  |--::/0,
          |--Caching: enabled,
          |--IPv4 resolution: enabled,
          |--IPv6 resolution: disabled,
          |--Verbosity level: 1/5,
          |--Verbosity details level: 0/4,
          |--Validation log level: 0/2,
          |--Blocked hostnames:,
          |--Blocked IP addresses:,
              |--127.0.0.1/8,
              |--10.0.0.0/8,
              |--172.16.0.0/12,
              |--192.168.0.0/16,
              |--169.254.0.0/16,
              |--::1/128,
              |--fc00::/7,
              |--fe80::/10,
              |--::ffff:0:0/96,
          |--Allowed hostnames:,
      |--Block malicious: enabled,
      |--Update: every 24h0m0s,
|--Firewall:,
|--System:,
   |--Process user ID: 1000,
   |--Process group ID: 1000,
   |--Timezone: america/new_york,
|--HTTP control server:,
   |--Listening port: 8000,
   |--Logging: enabled,
|--Public IP getter:,
   |--Fetch period: 12h0m0s,
   |--IP file: /tmp/gluetun/ip,
|--Github version information: enabled,
2021/04/17 10:24:31 INFO storage: merging by most recent 7378 hardcoded servers and 7378 servers read from /gluetun/servers.json,
2021/04/17 10:24:31 INFO routing: default route found: interface eth0, gateway 172.20.0.1,
2021/04/17 10:24:31 INFO routing: local ethernet link found: eth0,
2021/04/17 10:24:31 INFO routing: local ipnet found: 172.20.0.0/16,
2021/04/17 10:24:31 INFO routing: default route found: interface eth0, gateway 172.20.0.1,
2021/04/17 10:24:31 INFO routing: adding route for 0.0.0.0/0,
2021/04/17 10:24:31 INFO firewall: firewall disabled, only updating allowed subnets internal list,
2021/04/17 10:24:31 INFO routing: default route found: interface eth0, gateway 172.20.0.1,
2021/04/17 10:24:31 INFO openvpn configurator: checking for device /dev/net/tun,
2021/04/17 10:24:31 WARN TUN device is not available: open /dev/net/tun: no such file or directory,
2021/04/17 10:24:31 INFO openvpn configurator: creating /dev/net/tun,
2021/04/17 10:24:31 INFO firewall: enabling...,
2021/04/17 10:24:31 INFO firewall: enabled successfully,
2021/04/17 10:24:31 INFO healthcheck: listening on 127.0.0.1:9999,
2021/04/17 10:24:31 INFO dns over tls: using plaintext DNS at address 1.1.1.1,
2021/04/17 10:24:31 INFO http server: listening on 0.0.0.0:8000,
2021/04/17 10:24:31 INFO firewall: setting VPN connection through firewall...,
2021/04/17 10:24:31 INFO openvpn configurator: starting openvpn,
2021/04/17 10:24:31 INFO openvpn: OpenVPN 2.4.10 x86_64-alpine-linux-musl [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [MH/PKTINFO] [AEAD] built on Jan  4 2021,
2021/04/17 10:24:31 INFO openvpn: library versions: OpenSSL 1.1.1k  25 Mar 2021, LZO 2.10,
2021/04/17 10:24:31 INFO openvpn: CRL: loaded 1 CRLs from file [[INLINE]],
2021/04/17 10:24:31 INFO openvpn: TCP/UDP: Preserving recently used remote address: [AF_INET]154.3.40.40:1198,
2021/04/17 10:24:31 INFO openvpn: UDP link local: (not bound),
2021/04/17 10:24:31 INFO openvpn: UDP link remote: [AF_INET]154.3.40.40:1198,
2021/04/17 10:24:31 INFO openvpn: [toronto415] Peer Connection Initiated with [AF_INET]154.3.40.40:1198,
2021/04/17 10:24:33 INFO openvpn: OpenVPN ROUTE6: OpenVPN needs a gateway parameter for a --route-ipv6 option and no default was specified by either --route-ipv6-gateway or --ifconfig-ipv6 options,
2021/04/17 10:24:33 INFO openvpn: OpenVPN ROUTE: failed to parse/resolve route for host/network: 2000::/3,
2021/04/17 10:24:33 INFO openvpn: TUN/TAP device tun0 opened,
2021/04/17 10:24:33 INFO openvpn: /sbin/ip link set dev tun0 up mtu 1500,
2021/04/17 10:24:33 INFO openvpn: /sbin/ip addr add dev tun0 10.7.112.6/24 broadcast 10.7.112.255,
2021/04/17 10:24:33 WARN openvpn: OpenVPN was configured to add an IPv6 route over tun0. However, no IPv6 has been configured for this interface, therefore the route installation may fail or may not work as expected.,
2021/04/17 10:24:33 INFO openvpn: Initialization Sequence Completed,
2021/04/17 10:24:33 INFO VPN routing IP address: 154.3.40.40,
2021/04/17 10:24:33 INFO dns over tls: downloading DNS over TLS cryptographic files,
2021/04/17 10:24:33 INFO dns over tls: downloading hostnames and IP block lists,
2021/04/17 10:24:33 INFO healthcheck: healthy!,
2021/04/17 10:24:34 INFO dns over tls: init module 0: validator,
2021/04/17 10:24:34 INFO dns over tls: init module 1: iterator,
2021/04/17 10:24:34 INFO dns over tls: start of service (unbound 1.10.1).,
2021/04/17 10:24:34 INFO dns over tls: generate keytag query _ta-4a5c-4f66. NULL IN,
2021/04/17 10:24:34 INFO dns over tls: generate keytag query _ta-4a5c-4f66. NULL IN,
2021/04/17 10:24:35 INFO dns over tls: ready,
2021/04/17 10:24:35 INFO There is a new release v3.15.0 (v3.15.0) created 50 days ago,
2021/04/17 10:24:35 INFO VPN gateway IP address: 10.7.112.1,
panic: interface conversion: http.RoundTripper is nil, not *http.Transport,
,
goroutine 80 [running]:,
github.com/qdm12/gluetun/internal/provider.newPIAHTTPClient(0xc000106030, 0xa0fce8, 0xa, 0xc000095740, 0x0, 0x99a200),
	github.com/qdm12/gluetun/internal/provider/piav4.go:361 +0x3a6,
github.com/qdm12/gluetun/internal/provider.(*pia).PortForward(0xc0001620e0, 0xbce378, 0xc000326200, 0xc000106030, 0xc0000b4168, 0xbceff8, 0xc000100410, 0xc0000e3464, 0x4, 0x388, ...),
	github.com/qdm12/gluetun/internal/provider/piav4.go:224 +0x78,
github.com/qdm12/gluetun/internal/openvpn.(*looper).portForward(0xc0000d6a00, 0xbce378, 0xc000326200, 0xc0000a8540, 0xbcd170, 0xc0001620e0, 0xc000106030, 0xc0000e3464, 0x4, 0x388),
	github.com/qdm12/gluetun/internal/openvpn/loop.go:262 +0x1e7,
created by github.com/qdm12/gluetun/internal/openvpn.(*looper).Run.func1,
	github.com/qdm12/gluetun/internal/openvpn/loop.go:180 +0x105,
```

Notably, v3.15.0 works fine for me right now (brand new container that doesn't have an existing token).  

I hadn't restarted my containers in the last few weeks, so never encountered the same issues as @axxis1999.

--- a/Dockerfile
+++ b/Dockerfile
@@ -147,5 +147,4 @@ RUN apk add -q --progress --no-cache --update openvpn ca-certificates iptables i
     deluser unbound && \
     mkdir /gluetun
 # TODO remove once SAN is added to PIA servers certificates, see https://github.com/pia-foss/manual-connections/issues/10
-ENV GODEBUG=x509ignoreCN=0
 COPY --from=build /tmp/gobuild/entrypoint /entrypoint

--- a/internal/provider/piav4.go
+++ b/internal/provider/piav4.go
@@ -481,11 +481,13 @@ func fetchPIAToken(ctx context.Context, openFile os.OpenFileFunc,
 	}
 	request, err := http.NewRequestWithContext(ctx, http.MethodGet, url.String(), nil)
 	if err != nil {
-		return "", err
+		return "", replaceInErr(err, map[string]string{
+			username: "<username>", password: "<password>"})
 	}
 	response, err := client.Do(request)
 	if err != nil {
-		return "", err
+		return "", replaceInErr(err, map[string]string{
+			username: "<username>", password: "<password>"})
 	}
 	defer response.Body.Close()
 	if response.StatusCode != http.StatusOK {
@@ -541,10 +543,12 @@ func fetchPIAPortForwardData(ctx context.Context, client *http.Client, gateway n
 	}
 	request, err := http.NewRequestWithContext(ctx, http.MethodGet, url.String(), nil)
 	if err != nil {
+		err = replaceInErr(err, map[string]string{token: "<token>"})
 		return 0, "", expiration, fmt.Errorf("cannot obtain signature: %w", err)
 	}
 	response, err := client.Do(request)
 	if err != nil {
+		err = replaceInErr(err, map[string]string{token: "<token>"})
 		return 0, "", expiration, fmt.Errorf("cannot obtain signature: %w", err)
 	}
 	defer response.Body.Close()
@@ -584,11 +588,17 @@ func bindPIAPort(ctx context.Context, client *http.Client, gateway net.IP, data 
 
 	request, err := http.NewRequestWithContext(ctx, http.MethodGet, url.String(), nil)
 	if err != nil {
-		return fmt.Errorf("cannot bind port: %w", err)
+		return fmt.Errorf("cannot bind port: %w", replaceInErr(err, map[string]string{
+			payload:        "<payload>",
+			data.Signature: "<signature>",
+		}))
 	}
 	response, err := client.Do(request)
 	if err != nil {
-		return fmt.Errorf("cannot bind port: %w", err)
+		return fmt.Errorf("cannot bind port: %w", replaceInErr(err, map[string]string{
+			payload:        "<payload>",
+			data.Signature: "<signature>",
+		}))
 	}
 	defer response.Body.Close()
 	if response.StatusCode != http.StatusOK {
@@ -620,4 +630,13 @@ func writePortForwardedToFile(openFile os.OpenFileFunc,
 		return err
 	}
 	return file.Close()
+}
+
+// replaceInErr is used to remove sensitive information from logs.
+func replaceInErr(err error, substitutions map[string]string) error {
+	s := err.Error()
+	for old, new := range substitutions {
+		s = strings.ReplaceAll(s, old, new)
+	}
+	return errors.New(s)
 }

--- a/internal/provider/piav4.go
+++ b/internal/provider/piav4.go
@@ -358,8 +358,11 @@ func newPIAHTTPClient(parentClient *http.Client, serverName string) (client *htt
 		return nil, fmt.Errorf("cannot parse PIA root certificate: %w", err)
 	}
 
-	parentTransport := parentClient.Transport.(*http.Transport)
-	transport := parentTransport.Clone()
+	parentTransport := parentClient.Transport
+	if parentTransport == nil {
+		parentTransport = http.DefaultTransport
+	}
+	transport := parentTransport.(*http.Transport).Clone()
 
 	rootCAs := x509.NewCertPool()
 	rootCAs.AddCert(certificate)

--- a/internal/provider/piav4.go
+++ b/internal/provider/piav4.go
@@ -258,7 +258,10 @@ func (p *pia) PortForward(ctx context.Context, client *http.Client,
 
 	// First time binding
 	tryUntilSuccessful(ctx, pfLogger, func() error {
-		return bindPIAPort(ctx, privateIPClient, gateway, data)
+		if err := bindPIAPort(ctx, privateIPClient, gateway, data); err != nil {
+			return fmt.Errorf("cannot bind port: %w", err)
+		}
+		return nil
 	})
 	if ctx.Err() != nil {
 		return
@@ -295,7 +298,7 @@ func (p *pia) PortForward(ctx context.Context, client *http.Client,
 			return
 		case <-keepAliveTimer.C:
 			if err := bindPIAPort(ctx, privateIPClient, gateway, data); err != nil {
-				pfLogger.Error(err)
+				pfLogger.Error("cannot bind port: " + err.Error())
 			}
 			keepAliveTimer.Reset(keepAlivePeriod)
 		case <-expiryTimer.C:
@@ -323,7 +326,7 @@ func (p *pia) PortForward(ctx context.Context, client *http.Client,
 				pfLogger.Error(err)
 			}
 			if err := bindPIAPort(ctx, privateIPClient, gateway, data); err != nil {
-				pfLogger.Error(err)
+				pfLogger.Error("cannot bind port: " + err.Error())
 			}
 			if !keepAliveTimer.Stop() {
 				<-keepAliveTimer.C
@@ -588,17 +591,17 @@ func bindPIAPort(ctx context.Context, client *http.Client, gateway net.IP, data 
 
 	request, err := http.NewRequestWithContext(ctx, http.MethodGet, url.String(), nil)
 	if err != nil {
-		return fmt.Errorf("cannot bind port: %w", replaceInErr(err, map[string]string{
+		return replaceInErr(err, map[string]string{
 			payload:        "<payload>",
 			data.Signature: "<signature>",
-		}))
+		})
 	}
 	response, err := client.Do(request)
 	if err != nil {
-		return fmt.Errorf("cannot bind port: %w", replaceInErr(err, map[string]string{
+		return replaceInErr(err, map[string]string{
 			payload:        "<payload>",
 			data.Signature: "<signature>",
-		}))
+		})
 	}
 	defer response.Body.Close()
 	if response.StatusCode != http.StatusOK {
@@ -611,7 +614,7 @@ func bindPIAPort(ctx context.Context, client *http.Client, gateway net.IP, data 
 		Message string `json:"message"`
 	}
 	if err := decoder.Decode(&responseData); err != nil {
-		return fmt.Errorf("cannot bind port: %w", err)
+		return err
 	} else if responseData.Status != "OK" {
 		return fmt.Errorf("response received from PIA: %s (%s)", responseData.Status, responseData.Message)
 	}

--- a/internal/provider/piav4.go
+++ b/internal/provider/piav4.go
@@ -221,7 +221,7 @@ func (p *pia) PortForward(ctx context.Context, client *http.Client,
 		return
 	}
 
-	client, err := newPIAHTTPClient(client, commonName)
+	privateIPClient, err := newPIAHTTPClient(client, commonName)
 	if err != nil {
 		pfLogger.Error("aborting because: %s", err)
 		return
@@ -246,7 +246,7 @@ func (p *pia) PortForward(ctx context.Context, client *http.Client,
 
 	if !dataFound || expired {
 		tryUntilSuccessful(ctx, pfLogger, func() error {
-			data, err = refreshPIAPortForwardData(ctx, client, gateway, openFile)
+			data, err = refreshPIAPortForwardData(ctx, privateIPClient, gateway, openFile)
 			return err
 		})
 		if ctx.Err() != nil {
@@ -258,7 +258,7 @@ func (p *pia) PortForward(ctx context.Context, client *http.Client,
 
 	// First time binding
 	tryUntilSuccessful(ctx, pfLogger, func() error {
-		return bindPIAPort(ctx, client, gateway, data)
+		return bindPIAPort(ctx, privateIPClient, gateway, data)
 	})
 	if ctx.Err() != nil {
 		return
@@ -294,7 +294,7 @@ func (p *pia) PortForward(ctx context.Context, client *http.Client,
 			}
 			return
 		case <-keepAliveTimer.C:
-			if err := bindPIAPort(ctx, client, gateway, data); err != nil {
+			if err := bindPIAPort(ctx, privateIPClient, gateway, data); err != nil {
 				pfLogger.Error(err)
 			}
 			keepAliveTimer.Reset(keepAlivePeriod)
@@ -302,7 +302,7 @@ func (p *pia) PortForward(ctx context.Context, client *http.Client,
 			pfLogger.Warn("Forward port has expired on %s, getting another one", data.Expiration.Format(time.RFC1123))
 			oldPort := data.Port
 			for {
-				data, err = refreshPIAPortForwardData(ctx, client, gateway, openFile)
+				data, err = refreshPIAPortForwardData(ctx, privateIPClient, gateway, openFile)
 				if err != nil {
 					pfLogger.Error(err)
 					continue
@@ -322,7 +322,7 @@ func (p *pia) PortForward(ctx context.Context, client *http.Client,
 			if err := writePortForwardedToFile(openFile, filepath, data.Port); err != nil {
 				pfLogger.Error(err)
 			}
-			if err := bindPIAPort(ctx, client, gateway, data); err != nil {
+			if err := bindPIAPort(ctx, privateIPClient, gateway, data); err != nil {
 				pfLogger.Error(err)
 			}
 			if !keepAliveTimer.Stop() {

--- a/internal/provider/piav4.go
+++ b/internal/provider/piav4.go
@@ -246,7 +246,7 @@ func (p *pia) PortForward(ctx context.Context, client *http.Client,
 
 	if !dataFound || expired {
 		tryUntilSuccessful(ctx, pfLogger, func() error {
-			data, err = refreshPIAPortForwardData(ctx, privateIPClient, gateway, openFile)
+			data, err = refreshPIAPortForwardData(ctx, client, privateIPClient, gateway, openFile)
 			return err
 		})
 		if ctx.Err() != nil {
@@ -305,7 +305,7 @@ func (p *pia) PortForward(ctx context.Context, client *http.Client,
 			pfLogger.Warn("Forward port has expired on %s, getting another one", data.Expiration.Format(time.RFC1123))
 			oldPort := data.Port
 			for {
-				data, err = refreshPIAPortForwardData(ctx, privateIPClient, gateway, openFile)
+				data, err = refreshPIAPortForwardData(ctx, client, privateIPClient, gateway, openFile)
 				if err != nil {
 					pfLogger.Error(err)
 					continue
@@ -382,13 +382,13 @@ func newPIAHTTPClient(parentClient *http.Client, serverName string) (client *htt
 	}, nil
 }
 
-func refreshPIAPortForwardData(ctx context.Context, client *http.Client,
+func refreshPIAPortForwardData(ctx context.Context, client, privateIPClient *http.Client,
 	gateway net.IP, openFile os.OpenFileFunc) (data piaPortForwardData, err error) {
 	data.Token, err = fetchPIAToken(ctx, openFile, client)
 	if err != nil {
 		return data, fmt.Errorf("cannot obtain token: %w", err)
 	}
-	data.Port, data.Signature, data.Expiration, err = fetchPIAPortForwardData(ctx, client, gateway, data.Token)
+	data.Port, data.Signature, data.Expiration, err = fetchPIAPortForwardData(ctx, privateIPClient, gateway, data.Token)
 	if err != nil {
 		return data, fmt.Errorf("cannot obtain port forwarding data: %w", err)
 	}

--- a/internal/provider/piav4.go
+++ b/internal/provider/piav4.go
@@ -221,7 +221,7 @@ func (p *pia) PortForward(ctx context.Context, client *http.Client,
 		return
 	}
 
-	client, err := newPIAHTTPClient(commonName)
+	client, err := newPIAHTTPClient(client, commonName)
 	if err != nil {
 		pfLogger.Error("aborting because: %s", err)
 		return
@@ -348,7 +348,7 @@ func filterPIAServers(servers []models.PIAServer, regions []string, protocol str
 	return filtered
 }
 
-func newPIAHTTPClient(serverName string) (client *http.Client, err error) {
+func newPIAHTTPClient(parentClient *http.Client, serverName string) (client *http.Client, err error) {
 	certificateBytes, err := base64.StdEncoding.DecodeString(constants.PIACertificateStrong)
 	if err != nil {
 		return nil, fmt.Errorf("cannot decode PIA root certificate: %w", err)
@@ -357,32 +357,23 @@ func newPIAHTTPClient(serverName string) (client *http.Client, err error) {
 	if err != nil {
 		return nil, fmt.Errorf("cannot parse PIA root certificate: %w", err)
 	}
-	// certificate.DNSNames = []string{serverName, "10.0.0.1"}
+
+	parentTransport := parentClient.Transport.(*http.Transport)
+	transport := parentTransport.Clone()
+
 	rootCAs := x509.NewCertPool()
 	rootCAs.AddCert(certificate)
-	TLSClientConfig := &tls.Config{
+	transport.TLSClientConfig = &tls.Config{
 		RootCAs:    rootCAs,
 		MinVersion: tls.VersionTLS12,
 		ServerName: serverName,
 	}
-	//nolint:gomnd
-	transport := http.Transport{
-		TLSClientConfig: TLSClientConfig,
-		Proxy:           http.ProxyFromEnvironment,
-		DialContext: (&net.Dialer{
-			Timeout:   30 * time.Second,
-			KeepAlive: 30 * time.Second,
-			DualStack: true,
-		}).DialContext,
-		ForceAttemptHTTP2:     true,
-		MaxIdleConns:          100,
-		IdleConnTimeout:       90 * time.Second,
-		TLSHandshakeTimeout:   10 * time.Second,
-		ExpectContinueTimeout: 1 * time.Second,
-	}
+
 	const httpTimeout = 30 * time.Second
-	client = &http.Client{Transport: &transport, Timeout: httpTimeout}
-	return client, nil
+	return &http.Client{
+		Transport: transport,
+		Timeout:   httpTimeout,
+	}, nil
 }
 
 func refreshPIAPortForwardData(ctx context.Context, client *http.Client,

--- a/internal/provider/piav4.go
+++ b/internal/provider/piav4.go
@@ -482,8 +482,8 @@ func fetchPIAToken(ctx context.Context, openFile os.OpenFileFunc,
 	url := url.URL{
 		Scheme: "https",
 		User:   url.UserPassword(username, password),
-		Host:   "10.0.0.1",
-		Path:   "/authv3/generateToken",
+		Host:   "privateinternetaccess.com",
+		Path:   "/gtoken/generateToken",
 	}
 	request, err := http.NewRequestWithContext(ctx, http.MethodGet, url.String(), nil)
 	if err != nil {

--- a/internal/provider/piav4.go
+++ b/internal/provider/piav4.go
@@ -221,7 +221,7 @@ func (p *pia) PortForward(ctx context.Context, client *http.Client,
 		return
 	}
 
-	privateIPClient, err := newPIAHTTPClient(client, commonName)
+	privateIPClient, err := newPIAHTTPClient(commonName)
 	if err != nil {
 		pfLogger.Error("aborting because: %s", err)
 		return
@@ -351,7 +351,7 @@ func filterPIAServers(servers []models.PIAServer, regions []string, protocol str
 	return filtered
 }
 
-func newPIAHTTPClient(parentClient *http.Client, serverName string) (client *http.Client, err error) {
+func newPIAHTTPClient(serverName string) (client *http.Client, err error) {
 	certificateBytes, err := base64.StdEncoding.DecodeString(constants.PIACertificateStrong)
 	if err != nil {
 		return nil, fmt.Errorf("cannot decode PIA root certificate: %w", err)
@@ -361,12 +361,20 @@ func newPIAHTTPClient(parentClient *http.Client, serverName string) (client *htt
 		return nil, fmt.Errorf("cannot parse PIA root certificate: %w", err)
 	}
 
-	parentTransport := parentClient.Transport
-	if parentTransport == nil {
-		parentTransport = http.DefaultTransport
+	//nolint:gomnd
+	transport := &http.Transport{
+		// Settings taken from http.DefaultTransport
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+		}).DialContext,
+		ForceAttemptHTTP2:     true,
+		MaxIdleConns:          100,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
 	}
-	transport := parentTransport.(*http.Transport).Clone()
-
 	rootCAs := x509.NewCertPool()
 	rootCAs.AddCert(certificate)
 	transport.TLSClientConfig = &tls.Config{

--- a/internal/provider/piav4.go
+++ b/internal/provider/piav4.go
@@ -453,15 +453,15 @@ func writePIAPortForwardData(openFile os.OpenFileFunc, data piaPortForwardData) 
 }
 
 func unpackPIAPayload(payload string) (port uint16, token string, expiration time.Time, err error) {
-	b, err := base64.RawStdEncoding.DecodeString(payload)
+	b, err := base64.StdEncoding.DecodeString(payload)
 	if err != nil {
 		return 0, "", expiration,
-			fmt.Errorf("cannot decode payload: payload is %s: %w", payload, err)
+			fmt.Errorf("cannot decode payload: payload is %q: %w", payload, err)
 	}
 	var payloadData piaPayload
 	if err := json.Unmarshal(b, &payloadData); err != nil {
 		return 0, "", expiration,
-			fmt.Errorf("cannot parse payload data: data is %s: %w", string(b), err)
+			fmt.Errorf("cannot parse payload data: data is %q: %w", string(b), err)
 	}
 	return payloadData.Port, payloadData.Token, payloadData.Expiration, nil
 }
@@ -476,7 +476,7 @@ func packPIAPayload(port uint16, token string, expiration time.Time) (payload st
 	if err != nil {
 		return "", fmt.Errorf("cannot serialize payload data: %w", err)
 	}
-	payload = base64.RawStdEncoding.EncodeToString(b)
+	payload = base64.StdEncoding.EncodeToString(b)
 	return payload, nil
 }
 

--- a/internal/provider/piav4.go
+++ b/internal/provider/piav4.go
@@ -455,11 +455,13 @@ func writePIAPortForwardData(openFile os.OpenFileFunc, data piaPortForwardData) 
 func unpackPIAPayload(payload string) (port uint16, token string, expiration time.Time, err error) {
 	b, err := base64.RawStdEncoding.DecodeString(payload)
 	if err != nil {
-		return 0, "", expiration, fmt.Errorf("cannot decode payload: %w", err)
+		return 0, "", expiration,
+			fmt.Errorf("cannot decode payload: payload is %s: %w", payload, err)
 	}
 	var payloadData piaPayload
 	if err := json.Unmarshal(b, &payloadData); err != nil {
-		return 0, "", expiration, fmt.Errorf("cannot parse payload data: %w", err)
+		return 0, "", expiration,
+			fmt.Errorf("cannot parse payload data: data is %s: %w", string(b), err)
 	}
 	return payloadData.Port, payloadData.Token, payloadData.Expiration, nil
 }

--- a/internal/provider/piav4_test.go
+++ b/internal/provider/piav4_test.go
@@ -1,0 +1,45 @@
+package provider
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"net/http"
+	"testing"
+
+	"github.com/qdm12/gluetun/internal/constants"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func Test_newPIAHTTPClient(t *testing.T) {
+	t.Parallel()
+
+	const serverName = "testserver"
+
+	certificateBytes, err := base64.StdEncoding.DecodeString(constants.PIACertificateStrong)
+	require.NoError(t, err)
+	certificate, err := x509.ParseCertificate(certificateBytes)
+	require.NoError(t, err)
+	rootCAs := x509.NewCertPool()
+	rootCAs.AddCert(certificate)
+	expectedRootCAsSubjects := rootCAs.Subjects()
+
+	expectedPIATransportTLSConfig := &tls.Config{
+		// Can't directly compare RootCAs because of private fields
+		RootCAs:    nil,
+		MinVersion: tls.VersionTLS12,
+		ServerName: serverName,
+	}
+
+	piaClient, err := newPIAHTTPClient(serverName)
+
+	require.NoError(t, err)
+
+	// Verify pia transport TLS config is set
+	piaTransport := piaClient.Transport.(*http.Transport)
+	rootCAsSubjects := piaTransport.TLSClientConfig.RootCAs.Subjects()
+	assert.Equal(t, expectedRootCAsSubjects, rootCAsSubjects)
+	piaTransport.TLSClientConfig.RootCAs = nil
+	assert.Equal(t, expectedPIATransportTLSConfig, piaTransport.TLSClientConfig)
+}

--- a/internal/provider/piav4_test.go
+++ b/internal/provider/piav4_test.go
@@ -4,8 +4,11 @@ import (
 	"crypto/tls"
 	"crypto/x509"
 	"encoding/base64"
+	"encoding/json"
+	"errors"
 	"net/http"
 	"testing"
+	"time"
 
 	"github.com/qdm12/gluetun/internal/constants"
 	"github.com/stretchr/testify/assert"
@@ -42,4 +45,70 @@ func Test_newPIAHTTPClient(t *testing.T) {
 	assert.Equal(t, expectedRootCAsSubjects, rootCAsSubjects)
 	piaTransport.TLSClientConfig.RootCAs = nil
 	assert.Equal(t, expectedPIATransportTLSConfig, piaTransport.TLSClientConfig)
+}
+
+func Test_unpackPIAPayload(t *testing.T) {
+	t.Parallel()
+
+	const exampleToken = "token"
+	const examplePort = 2000
+	exampleExpiration := time.Unix(1000, 0).UTC()
+
+	testCases := map[string]struct {
+		payload    string
+		port       uint16
+		token      string
+		expiration time.Time
+		err        error
+	}{
+		"valid payload": {
+			payload:    makePIAPayload(t, exampleToken, examplePort, exampleExpiration),
+			port:       examplePort,
+			token:      exampleToken,
+			expiration: exampleExpiration,
+			err:        nil,
+		},
+		"invalid base64 payload": {
+			payload: "invalid",
+			err:     errors.New(`cannot decode payload: payload is "invalid": illegal base64 data at input byte 4`),
+		},
+		"invalid json payload": {
+			payload: base64.StdEncoding.EncodeToString([]byte{1}),
+			err:     errors.New(`cannot parse payload data: data is "\x01": invalid character '\x01' looking for beginning of value`), //nolint:lll
+		},
+	}
+
+	for name, testCase := range testCases {
+		testCase := testCase
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			port, token, expiration, err := unpackPIAPayload(testCase.payload)
+
+			if testCase.err != nil {
+				require.Error(t, err)
+				assert.Equal(t, testCase.err.Error(), err.Error())
+			} else {
+				require.NoError(t, err)
+			}
+
+			assert.Equal(t, testCase.port, port)
+			assert.Equal(t, testCase.token, token)
+			assert.Equal(t, testCase.expiration, expiration)
+		})
+	}
+}
+
+func makePIAPayload(t *testing.T, token string, port uint16, expiration time.Time) (payload string) {
+	t.Helper()
+
+	data := piaPayload{
+		Token:      token,
+		Port:       port,
+		Expiration: expiration,
+	}
+
+	b, err := json.Marshal(data)
+	require.NoError(t, err)
+
+	return base64.StdEncoding.EncodeToString(b)
 }

